notebooks/shapley_utility_learning.ipynb

```diff
@@ -0,0 +1,651 @@
+   "execution_count": 7,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "2022-10-18 22:12:37,742\tINFO worker.py:1509 -- Started a local Ray instance. View the dashboard at \u001b[1m\u001b[32mhttp://127.0.0.1:8265 \u001b[39m\u001b[22m\n"
+     ]
+    }
+   ],
+   "source": [
+    "df = compute_shapley_values(\n",
+    "    u=utility,\n",
+    "    mode=\"combinatorial_exact\",\n",
+    "    n_jobs=min(len(dataset), available_cpus()),\n",
+    "    progress=False,  # Does not display correctly in a notebook\n",
+    ")\n",
+    "df = df.rename(columns={\"data_value\": \"exact\"})"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We now estimate the Data Shapley values using the [DataUtilityLearning](../pydvl/utils/utility.rst#pydvl.utils.utility.DataUtilityLearning) wrapper. This class wraps a [Utility](../pydvl/utils/utility.rst#pydvl.utils.utility.Utility) and delegates calls to it, up until a given budget. Every call yields a utility sample which is saved under the hood for training of the given utility model. Once the budget is exhausted, `DataUtilityLearning` fits the model to the utility samples and all subsequent calls use the learned model to predict the wrapped utility instead of delegating to it.\n",
+    "\n",
+    "For the utility model we follow the paper and use a fully connected neural network. To train it we use a total of `training_budget` utility samples. We repeat this multiple times for each training budget."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 8,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "Doing 10 runs for each of 32 different training budgets.\n"
+     ]
+    }
+   ],
+   "source": [
+    "mlp_kwargs = dict(\n",
+    "    hidden_layer_sizes=(20, 10),\n",
+    "    activation=\"relu\",\n",
+    "    solver=\"adam\",\n",
+    "    learning_rate_init=0.001,\n",
+    "    batch_size=32,\n",
+    "    max_iter=800,\n",
+    ")\n",
+    "\n",
+    "print(\n",
+    "    f\"Doing {n_iterations} runs for each of {len(training_budget_values)} different training budgets.\"\n",
+    ")\n",
+    "\n",
+    "pbar = tqdm(\n",
+    "    product(range(n_iterations), training_budget_values),\n",
+    "    total=n_iterations * len(training_budget_values),\n",
+    ")\n",
+    "for idx, budget in pbar:\n",
+    "    pbar.set_postfix_str(f\"Run {idx} for training budget: {budget}\")\n",
+    "    dul_utility = DataUtilityLearning(\n",
+    "        u=utility, training_budget=budget, model=MLPRegressor(**mlp_kwargs)\n",
+    "    )\n",
+    "\n",
+    "    dul_df = compute_shapley_values(\n",
+    "        u=dul_utility,\n",
+    "        mode=\"truncated_montecarlo\",\n",
+    "        n_jobs=min(len(dataset), available_cpus()),\n",
+    "        progress=False,\n",
+    "        max_iterations=max(training_budget_values),\n",
```

Comment:
By setting `max_iterations=max(training_budget_values)` aren't we just reducing the effect of DUL as the training budget increases? Wouldn't be better to set it to a higher value, for e.g. `max_iterations= 2 * max(training_budget_values)` or even `max_iterations=2 ** len(u.data)`, to make sure that the only change happens because of DUL and not because truncated montecarlo shapley?

Reply:
You are of course right. My sleepy brain interpreted max_iterations as the training budget itself! 🤦🏽 Also, it doesn't make sense to use TMCS if using a hard limit on iterations

Reply:
OTOH, something is wrong with combinatorial MC...